### DISCUSSION TITLE
Github token lifetime?

### DISCUSSION THREAD

**Discussion by jherico** [community/community#25699] [Actions]
<p>I’m working on a workflow with a fairly long build and I’m seeing failures in code designed to interact with the PR (adding a comment with a link to the build result) at the end of the build.</p>
<p>I’m running the workflow on Mac and PC and both flows use the same python script set up with the same environment variables in order to upload the build to S3 and then post a comment on the PR.  </p>
<p>The Mac build takes about a half hour and completes fine.  The PC build takes about 1 hour and 15 minutes before it runs the script at which point the script throws an exception:</p>
<p>401 {“message”: “Bad credentials”, “documentation_url”: “<a href="https://developer.github.com/v3" rel="nofollow noopener">https://developer.github.com/v3</a>”}</p>
<p>If I alter the build so it skips some of the longer duration components (and thus completes in less than an hour) it runs fine.  So I’m guessing that the token created for the workflow only has a lifetime of about an hour?  Github workflows are allowed to run for up to 6 hours though, so it seems like the token lifetime should be at least that long. </p>

**Comment by boegel** [comment, marked as answer]
<p>Indeed, the token expires after 1 hour. From <a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/authenticating-with-the-github_token" rel="nofollow noopener">https://help.github.com/en/actions/automating-your-workflow-with-github-actions/authenticating-with-the-github_token</a>:</p>
<p>“The GITHUB_TOKEN secret is a GitHub App installation access token.”</p>
<p>and</p>
<p>“The installation access token expires after 60 minutes.”</p>

**Comment by MikeDombo** [comment]
<p>So it expires at an hour, is there any workaround for this? I too want a long-ish running task to then comment or create an issue, but I’m getting the 401 error. What options do I have?</p>

**Comment by Shchvova** [comment]
<p>Only workaround I found is to create <a href="https://github.com/settings/tokens/" rel="nofollow noopener">personal access token</a>, add it to secrets (say, <code>PAT</code>), and use <code>${{ secrets.PAT }}</code> instead:</p>
<pre><code>- name: Create Release
        id: create_release
        uses: actions/create-release@v1
        env:
          GITHUB_TOKEN: ${{ secrets.PAT }}
        with:
          tag_name: ${{ github.ref }}
          ....
</code></pre>
<p>But this is suboptimal, since it actions would be from your name, not from name of the bot.</p>

**Comment by francisfuzz** [comment]
<aside class="quote no-group" data-username="mikedombo" data-post="3" data-topic="16837" data-full="true">
<div class="title">
<div class="quote-controls"></div>
 mikedombo:</div>
<blockquote>
<p>So it expires at an hour, is there any workaround for this? I too want a long-ish running task to then comment or create an issue, but I’m getting the 401 error. What options do I have?</p>
</blockquote>
</aside>
<p><a class="mention" href="/mikedombo">@mikedombo</a> - That’s a great question. I think it’s worth mentioning that the <a href="https://docs.github.com/en/actions/configuring-and-managing-workflows/authenticating-with-the-github_token#about-the-github_token-secret">Authenticating with the GITHUB_TOKEN: About the GITHUB_TOKEN secret</a> article recently updated with this note:</p>
<blockquote>
<p>Before each job begins, GitHub fetches an installation access token for the job. The token expires when the job is finished.</p>
</blockquote>
<p>Given this documented behavior, you could create a workflow with many jobs, where each job cites the <code>GITHUB_TOKEN</code> in an environment variable. I’m wondering if that would help with what you’re looking to accomplish?</p>
